### ISSUE TITLE
When MeCabOovProviderPlugin is in use, a repeated Katakana word string does not get tokenized

### ISSUE THREAD

**Issue by azagniotov** [WorksApplications/Sudachi#216]
Hello Team,

I am not sure if this a bug or a misconfiguration of something, but when trying to tokenize a long string with a repeated Katakana word, let's say: `テスト` * 100 times (i.e.: `"テストテストテストテスト...."`), I get back a single Morpheme where the surface is the whole original query string. 

This issue is **not** observed with:
1. Repeated Hiragana word, e.g.:  `くよくよ` * 100 times
2. Repeated Kanji word, e.g.: `令和` * 100 times
3. A **short** string with repeated Katakana word, e.g.: `テスト` * 8 times 

I am using Sudachi 0.7.3 with the original `sudachi.json` and `char.def` that come bundled with the jar. I know that the repeated Katakana word issue is (probably) due to MeCabOovProviderPlugin, because when I removed the plugin entry from the `sudachi.json` config, the repeated long Katakana string got tokenized as expected.

**Comment by eiennohito** [comment]
TL;DR: it is a default behavior which you can turn off.

Sudachi concatenates katakana words (if the are more than 3 of them in a sequence) in post-processing using the `https://github.com/WorksApplications/Sudachi/blob/develop/src/main/java/com/worksap/nlp/sudachi/JoinKatakanaOovPlugin.java`

Default configuration: https://github.com/WorksApplications/Sudachi/blob/develop/src/main/resources/sudachi.json#L25C9-L28C10

You can remove the behavior by providing your own `sudachi.json` with `JoinKatakanaOovPlugin` removed.

**Comment by azagniotov** [comment]
@eiennohito thank you for your reply. 

Before making this post, I did try before to also remove the `JoinKatakanaOovPlugin`, unfortunately removing it does not fix this issue. 

Furthermore, as I mentioned in my original example, short (but greater than 3 words in sequence) Katakana strings do get tokenized. To put this in other words, when the `JoinKatakanaOovPlugin` is present with its default configuration as-is, the `テストテストテストテストテスト` (5 words in sequence) still gets tokenized into five morphemes:

<img width="867" alt="Screen Shot 2023-09-07 at 10 47 13 AM" src="https://github.com/WorksApplications/Sudachi/assets/989900/1ec55652-d9e5-4a4c-aa45-1a5e90d02323">

#### Understanding the minLength defined in the JSON

I debugged to understand how the `minLength` property defined in the JSON is treated and I see that basically it is used in the `JoinKatakanaOovPlugin`'s function:
```
   boolean isShorter(int length, InputText text, LatticeNode node) {
        return text.codePointCount(node.getBegin(), node.getEnd()) < length;
    }
```
Thus, it checks for the length of a word, not _how many words are there_ in sequence (I think this is what you alluded to in your reply). I validated my understanding by changing the `minLength` to be `4` instead of the default `3`, and indeed - the result was as expected - the `テストテストテストテストテスト` was not tokenized.

**But, this still does not solve my original question**
When `JoinKatakanaOovPlugin` is configured with its default configuration, why `テストテストテストテストテスト` gets tokenized (well, now we know - it is because of the `minLength` check), but the `テスト` * 100 does not?

#### Understanding what JoinKatakanaOovPlugin accepts to post-process

I did further debugging and this is what I see (please note the value of each `LatticeNodeImpl`) and the `path` size (how many nodes):
#### Breakpoint when the original input string contained (5) テスト in sequence

The `path` contains 5 nodes where each contains `テスト`
<img width="1029" alt="Screen Shot 2023-09-07 at 11 20 01 AM" src="https://github.com/WorksApplications/Sudachi/assets/989900/6068c8af-933f-4ffc-9121-64a661a34695">

#### Breakpoint when the original input string contained (100) テスト in sequence

The `path` contains 1 node where it contains the original un-tokenized `テスト` * 100
<img width="1031" alt="Screen Shot 2023-09-07 at 11 19 08 AM" src="https://github.com/WorksApplications/Sudachi/assets/989900/371a783e-4888-4937-87ec-a191d3ef45c8">


The above leads me to conclude that what `JoinKatakanaOovPlugin` receives to post-process is different due to reasons unrelated to `JoinKatakanaOovPlugin` itself: why in the former case the `path` contains five nodes, as opposed to the latter case when `path` contains a single node - should not it contain 100 nodes?


@eiennohito , I am sorry if I am asking the obvious questions 🙇🏼 . Please let me know how I can help more here to provide you with more information

**Comment by eiennohito** [comment]
Sorry for the confusion. I have double checked and there are several things playing around.

From some length (it seems "テスト" x 9), creating a single unknown word token for the whole string becomes less costlier from the analysis model point of view than creating a sequence of 9 テスト tokens.

Concatenating tokens in postprocessing stops playing a role in this time.
Still, I think this behavior is OK because 9 katakana tokens could not be reasonably segmented with Sudachi or any other MeCab-style morphological analyzers.

**Comment by eiennohito** [comment]
You are correct about `minLength` thing, it is about the minimum number of Unicode codepoints in the katakana word to begin concatenating them together, and it requires at least one token to be OOV in the sequence of katakana tokens

**Comment by azagniotov** [comment]
@eiennohito thank you for taking a deeper look! 

With regards to other MeCab-style tokenizers, I tested the `テスト` * 500 using the Kagome v2 tokenizer (and the latest UniDic), and I got 500 tokens as I expected.

To take a step back a little: I am not adamant about the need to tokenize enormous Katakana strings, I just wanted to raise this with you and the team to make sure that there is no bug lurking around.   

Based on what I am hearing from you, there is no bug and the behaviour I observed is as expected and nothing needs fixing, is this correct?

**Comment by eiennohito** [comment]
> With regards to other MeCab-style tokenizers, I tested the テスト * 500 using the Kagome v2 tokenizer (and the latest UniDic), and I got 500 tokens as I expected.

I am talking about the general case, not this (and pretty artifical) one (テスト * 500)
It still depends on the model itself and there could be some variations of course.

> Based on what I am hearing from you, there is no bug and the behaviour I observed is as expected and nothing needs fixing, is this correct?

Yes, current behavior is correct and expected for the present model.
For search... may be doing splits all of the time (as you expect) is better, not 100% sure though.
To fix it, the model should be re-learned from scratch and there is also a pile of worms going together with that thing.

**Comment by azagniotov** [comment]
Thank you for the discussion, I will close the current issue!

**Comment by eiennohito** [comment]
Feel free to ask if you have any questions
